Geometry/HGCalCommonData/src/HGCalGeomParameters.cc

```diff
@@ -688,118 +678,142 @@ void HGCalGeomParameters::loadGeometryHexagon8(const DDFilteredView& _fv, HGCalP
   std::map<int, HGCalGeomParameters::layerParameters> layers;
   std::map<std::pair<int, int>, HGCalParameters::hgtrform> trforms;
   int levelTop = 3 + std::max(php.levelT_[0], php.levelT_[1]);
+#ifdef EDM_ML_DEBUG
+  int ntot(0);
+#endif
   while (dodet) {
-    const DDSolid& sol = fv.logicalPart().solid();
-    // Layers first
+#ifdef EDM_ML_DEBUG
+    ++ntot;
+#endif
     std::vector<int> copy = fv.copyNumbers();
     int nsiz = (int)(copy.size());
-    int lay = (nsiz > levelTop) ? copy[nsiz - 4] : copy[nsiz - 1];
-    int zside = (nsiz > php.levelZSide_) ? copy[php.levelZSide_] : -1;
-    if (zside != 1)
-      zside = -1;
-    if (lay == 0) {
-      edm::LogError("HGCalGeom") << "Funny layer # " << lay << " zp " << zside << " in " << nsiz << " components";
-      throw cms::Exception("DDException") << "Funny layer # " << lay;
-    } else {
-      if (std::find(php.layer_.begin(), php.layer_.end(), lay) == php.layer_.end())
-        php.layer_.emplace_back(lay);
-      auto itr = layers.find(lay);
-      if (itr == layers.end()) {
-        const DDTubs& tube = static_cast<DDTubs>(sol);
-        double rin = HGCalParameters::k_ScaleFromDDD * tube.rIn();
-        double rout = HGCalParameters::k_ScaleFromDDD * tube.rOut();
-        double zp = HGCalParameters::k_ScaleFromDDD * fv.translation().Z();
-        HGCalGeomParameters::layerParameters laypar(rin, rout, zp);
-        layers[lay] = laypar;
-      }
-      if (trforms.find(std::make_pair(lay, zside)) == trforms.end()) {
-        DD3Vector x, y, z;
-        fv.rotation().GetComponents(x, y, z);
-        const CLHEP::HepRep3x3 rotation(x.X(), y.X(), z.X(), x.Y(), y.Y(), z.Y(), x.Z(), y.Z(), z.Z());
-        const CLHEP::HepRotation hr(rotation);
-        double xx =
-            ((std::abs(fv.translation().X()) < tolerance) ? 0 : HGCalParameters::k_ScaleFromDDD * fv.translation().X());
-        double yy =
-            ((std::abs(fv.translation().Y()) < tolerance) ? 0 : HGCalParameters::k_ScaleFromDDD * fv.translation().Y());
-        const CLHEP::Hep3Vector h3v(xx, yy, HGCalParameters::k_ScaleFromDDD * fv.translation().Z());
-        HGCalParameters::hgtrform mytrf;
-        mytrf.zp = zside;
-        mytrf.lay = lay;
-        mytrf.sec = 0;
-        mytrf.subsec = 0;
-        mytrf.h3v = h3v;
-        mytrf.hr = hr;
-        trforms[std::make_pair(lay, zside)] = mytrf;
+    if (nsiz < levelTop) {
+      int lay = copy[nsiz - 1];
+      int zside = (nsiz > php.levelZSide_) ? copy[php.levelZSide_] : -1;
+      if (zside != 1)
+        zside = -1;
+      const DDSolid& sol = fv.logicalPart().solid();
+#ifdef EDM_ML_DEBUG
+      edm::LogVerbatim("HGCalGeom") << sol.name() << " shape " << sol.shape() << " size " << nsiz << ":" << levelTop
+                                    << " lay " << lay << " z " << zside;
+#endif
+      if (lay == 0) {
+        edm::LogError("HGCalGeom") << "Funny layer # " << lay << " zp " << zside << " in " << nsiz << " components";
```

Comment:
no reason to emit LogError right before throwing an exception, just combine the two (the LogError message is more informative, so it should be used in the exception)

Reply:
Done